plugin/pulpcore/plugin/stages/artifact_stages.py

```diff
@@ -0,0 +1,274 @@
+                        if not incoming_content and not shutdown and not pending:
+                            content = await in_q.get()
+                            incoming_content.append(content)
+                            continue
+                    else:
+                        incoming_content.append(content)
+                        continue
+
+                for i, content in enumerate(incoming_content):
+                    if content is None:
+                        shutdown = True
+                        continue
+                    downloaders_for_content = []
+                    for declarative_artifact in content.d_artifacts:
+                        if declarative_artifact.artifact._state.adding:
+                            # this needs to be downloaded
+                            downloader = declarative_artifact.remote.get_downloader(
+                                declarative_artifact.url
+                            )
+                            next_future = asyncio.ensure_future(downloader.run())
+                            downloaders_for_content.append(next_future)
+                    if not downloaders_for_content:
+                        await out_q.put(content)
+                        continue
+
+                    async def return_content_for_downloader(c):
+                        return c
+                    outstanding_downloads = outstanding_downloads + len(downloaders_for_content)
+                    downloaders_for_content.append(return_content_for_downloader(content))
+                    pending.add(asyncio.gather(*downloaders_for_content))
+                    if outstanding_downloads > self.max_concurrent_downloads:
+                        saturated = True
+                        incoming_content = incoming_content[i + 1:]  # remove handled content
+                        break
+                else:
+                    incoming_content = []
+
+                if pending:
+                    done, pending = await asyncio.wait(pending, return_when=asyncio.FIRST_COMPLETED)
+                    for gathered_downloaders in done:
+                        results = gathered_downloaders.result()
+                        for download_result in results[:-1]:
+                            content = results[-1]
+                            for declarative_artifact in content.d_artifacts:
+                                if declarative_artifact.artifact._state.adding:
+                                    new_artifact = Artifact(
+                                        **download_result.artifact_attributes,
+                                        file=download_result.path
+                                    )
+                                    declarative_artifact.artifact = new_artifact
+                            pb.done = pb.done + len(content.d_artifacts)
+                            outstanding_downloads = outstanding_downloads - len(content.d_artifacts)
```

Comment:
Re:
```
 outstanding_downloads = outstanding_downloads - len(content.d_artifacts)
```
This seems to assume that all artifacts needed to be downloaded and had a downloader in `downloaders_for_content` which seems incorrect.

Reply:
Oh yeah I see what you are saying. I will fix this in the next push.